src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/ReferenceTrackerHost.cs

```diff
@@ -0,0 +1,209 @@
+        // Grid <---- NoCW             Grid <-------- NoCW
+        // | ^                         |              ^
+        // | |             Becomes     |              |
+        // v |                         v              |
+        // Rectangle                  Rectangle ----->Proxy
+        //
+        // Arguments
+        //   obj        - An IUnknown* where a NoCW points to (Grid, in this case)
+        //                    Notes:
+        //                    1. We can either create a new NoCW or get back an old one from the cache.
+        //                    2. This obj could be a regular tracker runtime object for data binding.
+        //  ppNewReference  - The IReferenceTrackerTarget* for the proxy created
+        //                    The tracker runtime will call IReferenceTrackerTarget to establish a reference.
+        //
+        [UnmanagedCallersOnly]
+        internal static unsafe int IReferenceTrackerHost_GetTrackerTarget(IntPtr pThis, IntPtr punk, IntPtr* ppNewReference)
+        {
+            if (punk == IntPtr.Zero)
+            {
+                return HResults.E_INVALIDARG;
+            }
+
+            if (Marshal.QueryInterface(punk, ComWrappers.IID_IUnknown, out IntPtr ppv) != HResults.S_OK)
+            {
+                return HResults.COR_E_INVALIDCAST;
+            }
+
+            try
+            {
+                using ComHolder identity = new ComHolder(ppv);
+                using ComHolder trackerTarget = new ComHolder(ComWrappers.GetOrCreateTrackerTarget(identity.Ptr));
+                return Marshal.QueryInterface(trackerTarget.Ptr, ComWrappers.IID_IReferenceTrackerTarget, out *ppNewReference);
+            }
+            catch (Exception e)
+            {
+                return Marshal.GetHRForException(e);
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        internal static unsafe int IReferenceTrackerHost_AddMemoryPressure(IntPtr pThis, long bytesAllocated)
+        {
+            try
+            {
+                GC.AddMemoryPressure(bytesAllocated);
+                return HResults.S_OK;
+            }
+            catch (Exception e)
+            {
+                return Marshal.GetHRForException(e);
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        internal static unsafe int IReferenceTrackerHost_RemoveMemoryPressure(IntPtr pThis, long bytesAllocated)
+        {
+            try
+            {
+                GC.RemoveMemoryPressure(bytesAllocated);
+                return HResults.S_OK;
+            }
+            catch (Exception e)
+            {
+                return Marshal.GetHRForException(e);
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        internal static unsafe int IReferenceTrackerHost_QueryInterface(IntPtr pThis, Guid* guid, IntPtr* ppObject)
+        {
+            if (*guid == ComWrappers.IID_IReferenceTrackerHost || *guid == ComWrappers.IID_IUnknown)
+            {
+                *ppObject = pThis;
```

Comment:
Do we need an `AddRef()` here? Callers will be calling `Release()`.

Reply:
We don't, as we use a custom IUnknown implementation that doesn't do any reference counting (see `CreateDefaultIReferenceTrackerHostVftbl`).

Reply:
Add a short comment that reference counting is not needed here?

Reply:
> We don't, as we use a custom IUnknown implementation that doesn't do any reference counting (see `CreateDefaultIReferenceTrackerHostVftbl`).

I wouldn't play this game. Just make the call and let the `AddRef` to the job it is supposed to. Making each of the COM APIs correct by construction is important. It costs us very little to do the very clear and correct thing, especially when they aren't in the same place.

`Marshal.AddRef(pThis);`